### PR TITLE
Cost endpoint fixes

### PR DESCRIPTION
**Pull request by tgautam23**
modified code with endpoint fixes

### CODE REVIEW

**Comment by dfridovi** [comment]
Please give this PR a meaningful title (can refer to other PRs for examples)

**Comment by dfridovi** [comment]
Also, not really a comment about the PR but just about the branch name, I think our convention has been either to (a) use <your-git-id>/<feature-name> or (b) <feature-type>/<feature-name>

File: exec/two_player_collision/main.cpp [hunk 1]
```diff
@@ -0,0 +1,251 @@
+/*
```
Comment (dfridovi) [inline exec/two_player_collision/main.cpp:1]: Also I'm looking at this again and realizing that actually it's probably simplest to keep this in two separate PRs like we talked about, i.e. one for the "endpoint fix" (which should not involved any new files) and a separate one introducing (and potentially modifying) the new example files. To that end, I'd remove this file (and the example header below) from this PR and move them to the other one which you'll open soon.

File: src/polyline2_signed_distance_constraint.cpp [hunk 2]
```diff
@@ -125,19 +128,6 @@ void Polyline2SignedDistanceConstraint::Quadraticize(const VectorXf& input,
...
-    const float grad_coeff = 2.0 * orientation * sign / level;
-    const float weighted_grad_coeff = weight_ * grad_coeff;
-    (*grad)(xidx_) += weighted_grad_coeff * dx;
-    (*grad)(yidx_) += weighted_grad_coeff * dy;
-
-    (*hess)(xidx_, xidx_) += weighted_grad_coeff * (grad_coeff * dx2 + 1.0);
-    (*hess)(yidx_, yidx_) += weighted_grad_coeff * (grad_coeff * dy2 + 1.0);
-
-    const float hess_xy = weighted_grad_coeff * grad_coeff * dx * dy;
-    (*hess)(xidx_, yidx_) += hess_xy;
```
Comment (dfridovi) [inline src/polyline2_signed_distance_constraint.cpp:139]: Wait why did this get deleted?

File: src/semiquadratic_polyline2_cost.cpp [hunk 3]
```diff
@@ -113,17 +117,6 @@ void SemiquadraticPolyline2Cost::Quadraticize(const VectorXf& input,
...
-    // Closest point is a vertex.
-    (*hess)(xidx_, xidx_) += weight_;
-    (*hess)(yidx_, yidx_) += weight_;
-
-    float scaling = std::sqrt(std::abs(signed_squared_distance));
-    scaling = (scaling - std::abs(threshold_)) / scaling;
-    (*grad)(xidx_) +=
-        weight_ * scaling * (current_position.x() - closest_point.x());
-    (*grad)(yidx_) +=
-        weight_ * scaling * (current_position.y() - closest_point.y());
```
Comment (dfridovi) [inline src/semiquadratic_polyline2_cost.cpp:126]: Why did this get deleted?

File: exec/two_player_collision/main.cpp (hunk 1 above)
Comment (tgautam23) [reply exec/two_player_collision/main.cpp]: I removed all the files relevant for the collision example and put them in a new branch. The files I moved are src/two_player_collision_example.cpp, the two_player_collision_example.h file in the include folder, and the exec/two_player_collision folder.

Comment (dfridovi) [reply exec/two_player_collision/main.cpp]: I left this in the other PR but it's relevant here, so I'll repeat. Anyway, I think you need to `git rm` the file to remove it from the commit itself. Or maybe it's enough to just `rm` the file and then do `git add -A` to make sure git registers the deleted file. This is causing your merge conflict in the other PR.

File: src/polyline2_signed_distance_constraint.cpp (hunk 2 above)
Comment (dfridovi) [reply src/polyline2_signed_distance_constraint.cpp:139]: ok I see this was an attempt to fix the case where closest point is a vertex. Unfortunately, this is going to break the case (which this was handling) where the closest point is indeed a vertex of the polyline but not an endpoint. To handle this correctly, I think you will need to edit the `ClosestPoint` function to set an `is_endpoint` flag when a returned closest point is actually an endpoint of the polyline.

File: src/quadratic_polyline2_cost.cpp [hunk 4]
```diff
@@ -55,9 +55,13 @@ float QuadraticPolyline2Cost::Evaluate(const VectorXf& input) const {
...
   // Compute signed squared distance by finding closest point.
   float signed_squared_distance;
-  polyline_.ClosestPoint(Point2(input(xidx_), input(yidx_)), nullptr, nullptr,
+  bool is_vertex;
+  polyline_.ClosestPoint(Point2(input(xidx_), input(yidx_)), &is_vertex, nullptr,
                          &signed_squared_distance);
-
+  if(is_vertex){
+    // endpoint fix
+  	signed_squared_distance = 0;
```
[the remark this answers is not in the document]
Comment (tgautam23) [reply src/quadratic_polyline2_cost.cpp:63]: The polyline2 closestpoint member function has been modified to include an endpoint boolean. Therefore, the above has now been modified to if(is_endpoint)...

File: src/quadratic_polyline2_cost.cpp [hunk 5]
```diff
@@ -101,11 +105,8 @@ void QuadraticPolyline2Cost::Quadraticize(const VectorXf& input, MatrixXf* hess,
     (*grad)(xidx_) += w_cross * unit_segment.y();
     (*grad)(yidx_) -= w_cross * unit_segment.x();
   } else {
-    (*hess)(xidx_, xidx_) += weight_;
-    (*hess)(yidx_, yidx_) += weight_;
-
-    (*grad)(xidx_) += weight_ * (current_position.x() - closest_point.x());
-    (*grad)(yidx_) += weight_ * (current_position.y() - closest_point.y());
+    // endpoint fix - when cost is 0 
+    return;
```
[the remark this answers is not in the document]
Comment (tgautam23) [reply src/quadratic_polyline2_cost.cpp:109]: The endpoint fix has now been implemented in the Quadraticize() member function. The fix treats endpoint and vertex cases separately.

File: src/quadratic_polyline2_cost.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dfridovi) [reply src/quadratic_polyline2_cost.cpp:63]: The `polyline2.h` and `.cpp` files do not appear to be modified in this branch. Perhaps they're in some other branch?

[the remark this answers is not in the document]
Comment (tgautam23) [reply src/quadratic_polyline2_cost.cpp:63]: Still adding the endpoint boolean to the PointAt() member function. Will push the files asap.

[the remark this answers is not in the document]
Comment (tgautam23) [reply src/quadratic_polyline2_cost.cpp:63]: Modified the files and pushed them to this branch.

File: src/route_progress_cost.cpp [hunk 6]
```diff
@@ -80,7 +83,10 @@ void RouteProgressCost::Quadraticize(Time t, const VectorXf& input,
   const float desired_route_pos =
       initial_route_pos_ + (t - initial_time_) * nominal_speed_;
 
-  const Point2 route_point = polyline_.PointAt(desired_route_pos);
+  bool is_vertex;
+  const Point2 route_point = polyline_.PointAt(desired_route_pos, &is_vertex, nullptr);
+  if(is_vertex) //endpoint fix
+    return;
```
[the remark this answers is not in the document]
Comment (tgautam23) [reply src/route_progress_cost.cpp:89]: Changed this to the is_endpoint flag.

File: src/semiquadratic_polyline2_cost.cpp [hunk 7]
```diff
@@ -55,9 +55,13 @@ float SemiquadraticPolyline2Cost::Evaluate(const VectorXf& input) const {
...
   // Compute signed squared distance by finding closest point.
   float signed_squared_distance;
-  polyline_.ClosestPoint(Point2(input(xidx_), input(yidx_)), nullptr, nullptr,
+  bool is_vertex;
+  polyline_.ClosestPoint(Point2(input(xidx_), input(yidx_)), &is_vertex, nullptr,
                          &signed_squared_distance);
-
+  if (is_vertex){
+    // endpoint fix
+    signed_squared_distance = 0;
```
[the remark this answers is not in the document]
Comment (tgautam23) [reply src/semiquadratic_polyline2_cost.cpp]: Changed this to a return.

File: src/polyline2_signed_distance_constraint.cpp (hunk 2 above)
Comment (tgautam23) [reply src/polyline2_signed_distance_constraint.cpp:139]: I put back all the cases where the closest point is a vertex. I implemented an if condition on the outside, first checking whether we are dealing with an endpoint.

File: include/ilqgames/examples/two_player_collision_example.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+ * Authors: David Fridovich-Keil   ( dfk@eecs.berkeley.edu )
+ */
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Two player collision example.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef ILQGAMES_EXAMPLE_TWO_PLAYER_COLLISION_EXAMPLE_H
```
[the remark this answers is not in the document]
Comment (tgautam23) [reply include/ilqgames/examples/two_player_collision_example.h]: are the two names not the same?

[the remark this answers is not in the document]
Comment (dfridovi) [reply include/ilqgames/examples/two_player_collision_example.h]: Well that's why I called it a nitpick... actually the "EXAMPLES" is plural in my comment.

[the remark this answers is not in the document]
Comment (tgautam23) [reply include/ilqgames/examples/two_player_collision_example.h]: fixed!

File: src/two_player_collision_example.cpp [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+ */
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Three player overtaking example. Ordering is given by the following:
+// (P1, P2, P3) = (Car 1, Car 2, Pedestrian).
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include <ilqgames/cost/curvature_cost.h>
```
Comment (dfridovi) [inline src/two_player_collision_example.cpp]: Why is this file here?

File: include/ilqgames/examples/two_player_collision_example.h [hunk 10]
```diff
@@ -0,0 +1,65 @@
+/*
```
Comment (dfridovi) [inline include/ilqgames/examples/two_player_collision_example.h]: Shouldn't this file be in the other PR?

File: exec/two_player_collision/main.cpp [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+ * Copyright (c) 2019, The Regents of the University of California (Regents).
```
Comment (dfridovi) [inline exec/two_player_collision/main.cpp]: This file should be in the other PR, right?

File: src/polyline2_signed_distance_constraint.cpp (hunk 2 above)
Comment (dfridovi) [reply src/polyline2_signed_distance_constraint.cpp:139]: Ok no I see this now. But my other comment still stands - i.e. it looks like you're actually checking if the closest point is a vertex, not an endpoint.

File: src/polyline2_signed_distance_constraint.cpp [hunk 12]
```diff
@@ -127,17 +130,7 @@ void Polyline2SignedDistanceConstraint::Quadraticize(const VectorXf& input,
...
-    (*hess)(yidx_, xidx_) += hess_xy;
+    return;
```
[the remark this answers is not in the document]
Comment (dfridovi) [reply src/polyline2_signed_distance_constraint.cpp]: Ping.

File: src/quadratic_polyline2_cost.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dfridovi) [reply src/quadratic_polyline2_cost.cpp:63]: They don't appear here. Are you sure they're pushed.

File: src/quadratic_polyline2_cost.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (dfridovi) [reply src/quadratic_polyline2_cost.cpp]: Nonononono there should be a separate Hessian that's not zero when the closest point is a vertex, and it should be zero when it's an endpoint.

File: src/polyline2.cpp [hunk 13]
```diff
@@ -89,33 +89,75 @@ Point2 Polyline2::PointAt(float route_pos, bool* is_vertex,
...
     }
+    counter += 1; 
   }
 
+  // check if the closest point occurs at an endpoint for the polyline
+  if (is_endpoint){
+    if(segment_ind == 0){
+      bool check_vertex;
+      float ssd;
+      segments_[segment_ind + 1].ClosestPoint(query, &check_vertex, &ssd);
```
Comment (dfridovi) [inline src/polyline2.cpp:142]: I think this is a little complicated logically, so maybe explain in a comment.

File: src/polyline2.cpp [hunk 14]
```diff
@@ -89,33 +89,75 @@ Point2 Polyline2::PointAt(float route_pos, bool* is_vertex,
...
 
-      if (is_vertex) *is_vertex = is_endpoint;
+      if (is_vertex) *is_vertex = is_segment_endpoint;
       if (segment) *segment = s;
+      segment_ind = counter;
     }
+    counter += 1; 
   }
 
+  // check if the closest point occurs at an endpoint for the polyline
```
Comment (dfridovi) [inline src/polyline2.cpp:137]: nit: capitalize first letter and add a period.

File: src/polyline2.cpp [hunk 15]
```diff
@@ -89,33 +89,75 @@ Point2 Polyline2::PointAt(float route_pos, bool* is_vertex,
...
+    }
+    else if(segment_ind == segments_.size()-1){
+      bool check_vertex;
+      float ssd;
+      segments_[segment_ind - 1].ClosestPoint(query, &check_vertex, &ssd);
+      if(!check_vertex){
+        *is_endpoint = true;
+      }
+    }
+    else{
```
Comment (dfridovi) [inline src/polyline2.cpp:155]: nit: you don't need braces on single liners like this, and you didn't use them above, so I'd probably try to stay consistent.

File: src/polyline2_signed_distance_constraint.cpp [hunk 16]
```diff
@@ -58,9 +58,12 @@ bool Polyline2SignedDistanceConstraint::IsSatisfied(const VectorXf& input,
 
   // Compute signed squared distance by finding closest point.
   float signed_distance_sq;
+  bool is_endpoint;
   polyline_.ClosestPoint(Point2(input(xidx_), input(yidx_)), nullptr, nullptr,
-                         &signed_distance_sq);
-
+                         &signed_distance_sq, &is_endpoint);
+  if(is_endpoint){
+    signed_distance_sq = 0;
```
Comment (dfridovi) [inline src/polyline2_signed_distance_constraint.cpp:65]: nit: 0.0

Comment (dfridovi) [reply src/polyline2_signed_distance_constraint.cpp:65]: there is actually a reason for this - by specifying 0.0 you make it unambiguous that the compiler should treat this as a floating point constant and not an integer

File: src/polyline2_signed_distance_constraint.cpp [hunk 17]
```diff
@@ -58,9 +58,12 @@ bool Polyline2SignedDistanceConstraint::IsSatisfied(const VectorXf& input,
...
+  }
   // Maybe set level.
```
Comment (dfridovi) [inline src/polyline2_signed_distance_constraint.cpp:67]: nit: blank line above this

File: src/polyline2_signed_distance_constraint.cpp [hunk 18]
```diff
@@ -102,30 +106,32 @@ void Polyline2SignedDistanceConstraint::Quadraticize(const VectorXf& input,
...
-    const float hess_xy = hess_coeff * unit_direction.x() * unit_direction.y();
-    (*hess)(xidx_, xidx_) +=
-        hess_coeff * unit_direction.y() * unit_direction.y();
-    (*hess)(yidx_, yidx_) +=
-        hess_coeff * unit_direction.x() * unit_direction.x();
-    (*hess)(xidx_, yidx_) -= hess_xy;
-    (*hess)(yidx_, xidx_) -= hess_xy;
-  } else {
+  // First check if closest point is an endpoint of the polyline
+  if(!is_endpoint){
```
Comment (dfridovi) [inline src/polyline2_signed_distance_constraint.cpp:110]: nit: spaces before and after parens, i.e. `if (!is_endpoint) {`

File: src/polyline2_signed_distance_constraint.cpp [hunk 19]
```diff
@@ -102,30 +106,32 @@ void Polyline2SignedDistanceConstraint::Quadraticize(const VectorXf& input,
...
+      const float hess_coeff = weight_ * coeff * (grad_coeff * cross + 1.0);
+      const float hess_xy = hess_coeff * unit_direction.x() * unit_direction.y();
+      (*hess)(xidx_, xidx_) +=
+          hess_coeff * unit_direction.y() * unit_direction.y();
+      (*hess)(yidx_, yidx_) +=
+          hess_coeff * unit_direction.x() * unit_direction.x();
+      (*hess)(xidx_, yidx_) -= hess_xy;
+      (*hess)(yidx_, xidx_) -= hess_xy;
+    } else {
     // Closest point is a vertex.
```
Comment (dfridovi) [inline src/polyline2_signed_distance_constraint.cpp:135]: indent

File: src/polyline2_signed_distance_constraint.cpp [hunk 20]
```diff
@@ -138,7 +144,8 @@ void Polyline2SignedDistanceConstraint::Quadraticize(const VectorXf& input,
     const float hess_xy = weighted_grad_coeff * grad_coeff * dx * dy;
     (*hess)(xidx_, yidx_) += hess_xy;
     (*hess)(yidx_, xidx_) += hess_xy;
-  }
+    }
+  } 
 }
 
 }  // namespace ilqgames
```
Comment (dfridovi) [inline src/polyline2_signed_distance_constraint.cpp:151]: nice fix to this file

File: src/quadratic_polyline2_cost.cpp [hunk 21]
```diff
@@ -55,9 +55,13 @@ float QuadraticPolyline2Cost::Evaluate(const VectorXf& input) const {
...
   // Compute signed squared distance by finding closest point.
   float signed_squared_distance;
+  bool is_endpoint;
   polyline_.ClosestPoint(Point2(input(xidx_), input(yidx_)), nullptr, nullptr,
-                         &signed_squared_distance);
-
+                         &signed_squared_distance, &is_endpoint);
+  if(is_endpoint){
+    // endpoint fix
+  	signed_squared_distance = 0;
```
Comment (dfridovi) [inline src/quadratic_polyline2_cost.cpp:63]: nit: see above about setting this to 0.0

File: src/quadratic_polyline2_cost.cpp [hunk 22]
```diff
@@ -55,9 +55,13 @@ float QuadraticPolyline2Cost::Evaluate(const VectorXf& input) const {
...
+  }
   return 0.5 * weight_ * std::abs(signed_squared_distance);
```
Comment (dfridovi) [inline src/quadratic_polyline2_cost.cpp:65]: nit: blank line before

File: src/quadratic_polyline2_cost.cpp [hunk 23]
```diff
@@ -76,37 +80,42 @@ void QuadraticPolyline2Cost::Quadraticize(const VectorXf& input, MatrixXf* hess,
...
 
   bool is_vertex;
+  bool is_endpoint;
   LineSegment2 segment(Point2(0.0, 0.0), Point2(1.0, 1.0));
   const Point2 closest_point =
-      polyline_.ClosestPoint(current_position, &is_vertex, &segment, nullptr);
+      polyline_.ClosestPoint(current_position, &is_vertex, &segment, nullptr, &is_endpoint);
 
+  //First check whether the closest point is a endpoint of the polyline.
+  if(!is_endpoint){
```
Comment (dfridovi) [inline src/quadratic_polyline2_cost.cpp:89]: nit: whitespace

File: src/quadratic_polyline2_cost.cpp [hunk 24]
```diff
@@ -76,37 +80,42 @@ void QuadraticPolyline2Cost::Quadraticize(const VectorXf& input, MatrixXf* hess,
...
+      (*grad)(yidx_) -= w_cross * unit_segment.x();
+    } else {
+      (*hess)(xidx_, xidx_) += weight_;
+      (*hess)(yidx_, yidx_) += weight_;
+
+      (*grad)(xidx_) += weight_ * (current_position.x() - closest_point.x());
+      (*grad)(yidx_) += weight_ * (current_position.y() - closest_point.y());   
+    }
+  } else
+    return; // endpoint fix - when cost is 0
```
Comment (dfridovi) [inline src/quadratic_polyline2_cost.cpp:118]: just nix the `else return` since it's unnecessary

File: src/route_progress_cost.cpp [hunk 25]
```diff
@@ -56,8 +56,11 @@ float RouteProgressCost::Evaluate(Time t, const VectorXf& input) const {
 
   const float desired_route_pos =
       initial_route_pos_ + (t - initial_time_) * nominal_speed_;
-
-  const Point2 desired = polyline_.PointAt(desired_route_pos);
+  bool is_endpoint;
+  const Point2 desired = polyline_.PointAt(desired_route_pos, nullptr, nullptr, &is_endpoint);
+  if(is_endpoint){// endpoint fix
```
Comment (dfridovi) [inline src/route_progress_cost.cpp:61]: see earlier comments about whitespace and 0.0

File: src/semiquadratic_polyline2_cost.cpp [hunk 26]
```diff
@@ -55,9 +55,13 @@ float SemiquadraticPolyline2Cost::Evaluate(const VectorXf& input) const {
...
   // Compute signed squared distance by finding closest point.
   float signed_squared_distance;
+  bool is_endpoint;
   polyline_.ClosestPoint(Point2(input(xidx_), input(yidx_)), nullptr, nullptr,
-                         &signed_squared_distance);
-
+                         &signed_squared_distance, &is_endpoint);
+  if (is_endpoint){
+    // endpoint fix
+    return 0;
```
Comment (dfridovi) [inline src/semiquadratic_polyline2_cost.cpp:63]: 0.0

File: src/polyline2.cpp [hunk 27]
```diff
@@ -89,33 +89,75 @@ Point2 Polyline2::PointAt(float route_pos, bool* is_vertex,
                  remaining > segments_[idx].Length();
   }
 
-  return segments_[idx].FirstPoint() +
+  Point2 return_point = segments_[idx].FirstPoint() +
```
[the remark this answers is not in the document]
Comment (tgautam23) [reply src/polyline2.cpp:92]: done!

File: src/polyline2.cpp [hunk 28]
```diff
@@ -89,33 +89,75 @@ Point2 Polyline2::PointAt(float route_pos, bool* is_vertex,
...
          remaining * segments_[idx].UnitDirection();
+  if (is_endpoint) {
+    if(idx == 0){
```
[the remark this answers is not in the document]
Comment (tgautam23) [reply src/polyline2.cpp:95]: done!

File: src/polyline2.cpp [hunk 29]
```diff
@@ -89,33 +89,75 @@ Point2 Polyline2::PointAt(float route_pos, bool* is_vertex,
                  remaining > segments_[idx].Length();
   }
 
-  return segments_[idx].FirstPoint() +
+  Point2 return_point = segments_[idx].FirstPoint() +
          remaining * segments_[idx].UnitDirection();
+  if (is_endpoint) {
```
[the remark this answers is not in the document]
Comment (tgautam23) [reply src/polyline2.cpp:94]: done!

File: src/polyline2.cpp [hunk 30]
```diff
@@ -89,33 +89,75 @@ Point2 Polyline2::PointAt(float route_pos, bool* is_vertex,
...
                                LineSegment2* segment,
-                               float* signed_squared_distance) const {
+                               float* signed_squared_distance, bool* is_endpoint) const {
   // Walk along each line segment and remember which was closest.
   float closest_signed_squared_distance = constants::kInfinity;
   Point2 closest_point;
 
   float current_signed_squared_distance;
+  int segment_ind = 0;
+  int counter = 0;
```
[the remark this answers is not in the document]
Comment (tgautam23) [reply src/polyline2.cpp:119]: done!

File: src/polyline2.cpp [hunk 31]
```diff
@@ -89,33 +89,75 @@ Point2 Polyline2::PointAt(float route_pos, bool* is_vertex,
...
         std::abs(closest_signed_squared_distance)) {
       closest_signed_squared_distance = current_signed_squared_distance;
       closest_point = current_point;
 
-      if (is_vertex) *is_vertex = is_endpoint;
+      if (is_vertex) *is_vertex = is_segment_endpoint;
       if (segment) *segment = s;
+      segment_ind = counter;
     }
+    counter += 1;
```
[the remark this answers is not in the document]
Comment (tgautam23) [reply src/polyline2.cpp:134]: done!

File: src/polyline2.cpp (hunk 13 above)
Comment (tgautam23) [reply src/polyline2.cpp:142]: done!

File: src/polyline2.cpp (hunk 14 above)
Comment (tgautam23) [reply src/polyline2.cpp:137]: Changed comments to this formatting style.

File: src/polyline2.cpp (hunk 15 above)
Comment (tgautam23) [reply src/polyline2.cpp:155]: removed braces.

File: src/polyline2_signed_distance_constraint.cpp (hunk 16 above)
Comment (tgautam23) [reply src/polyline2_signed_distance_constraint.cpp:65]: done!

File: src/polyline2_signed_distance_constraint.cpp (hunk 17 above)
Comment (tgautam23) [reply src/polyline2_signed_distance_constraint.cpp:67]: done!

File: src/polyline2_signed_distance_constraint.cpp (hunk 18 above)
Comment (tgautam23) [reply src/polyline2_signed_distance_constraint.cpp:110]: done!

File: src/polyline2_signed_distance_constraint.cpp (hunk 19 above)
Comment (tgautam23) [reply src/polyline2_signed_distance_constraint.cpp:135]: done!

File: src/polyline2_signed_distance_constraint.cpp (hunk 20 above)
Comment (tgautam23) [reply src/polyline2_signed_distance_constraint.cpp:151]: thanks :)

File: src/quadratic_polyline2_cost.cpp (hunk 21 above)
Comment (tgautam23) [reply src/quadratic_polyline2_cost.cpp:63]: done!

File: src/quadratic_polyline2_cost.cpp (hunk 22 above)
Comment (tgautam23) [reply src/quadratic_polyline2_cost.cpp:65]: done!

File: src/quadratic_polyline2_cost.cpp (hunk 23 above)
Comment (tgautam23) [reply src/quadratic_polyline2_cost.cpp:89]: done!

File: src/quadratic_polyline2_cost.cpp (hunk 24 above)
Comment (tgautam23) [reply src/quadratic_polyline2_cost.cpp:118]: removed.

File: src/route_progress_cost.cpp (hunk 25 above)
Comment (tgautam23) [reply src/route_progress_cost.cpp:61]: done!

File: src/semiquadratic_polyline2_cost.cpp (hunk 26 above)
Comment (tgautam23) [reply src/semiquadratic_polyline2_cost.cpp:63]: done!

File: include/ilqgames/geometry/polyline2.h [hunk 32]
```diff
@@ -68,13 +68,13 @@ class Polyline2 {
   // vertex), and the signed squared distance, where right is positive.
   Point2 ClosestPoint(const Point2& query, bool* is_vertex = nullptr,
                       LineSegment2* segment = nullptr,
-                      float* signed_squared_distance = nullptr) const;
+                      float* signed_squared_distance = nullptr, bool* is_endpoint = nullptr) const;
```
[the remark this answers is not in the document]
Comment (dfridovi) [reply include/ilqgames/geometry/polyline2.h:71]: If you're not going to do this in this PR, please do post an issue asking that it be done. Someone (I, Vince, you, etc) should get to it soon.

File: src/polyline2.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (dfridovi) [reply src/polyline2.cpp:94]: Uhh... this and other changes don't show up. Please push any local changes.

File: src/polyline2.cpp (hunk 13 above)
Comment (dfridovi) [reply src/polyline2.cpp:145]: Nice! Also, can you please get rid of the `float ssd` since the `ClosestPoint()` function just defaults that argument to `nullptr` anyway and you don't use the value here.

File: src/polyline2.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (dfridovi) [reply src/polyline2.cpp:94]: ping....

File: src/polyline2_signed_distance_constraint.cpp (hunk 17 above)
Comment (dfridovi) [reply src/polyline2_signed_distance_constraint.cpp:68]: See comment above about not seeing change here.

File: src/quadratic_polyline2_cost.cpp (hunk 22 above)
Comment (dfridovi) [reply src/quadratic_polyline2_cost.cpp:66]: See comment above about not seeing change.

File: src/semiquadratic_polyline2_cost.cpp [hunk 33]
```diff
@@ -55,9 +55,13 @@ float SemiquadraticPolyline2Cost::Evaluate(const VectorXf& input) const {
 
   // Compute signed squared distance by finding closest point.
   float signed_squared_distance;
+  bool is_endpoint;
   polyline_.ClosestPoint(Point2(input(xidx_), input(yidx_)), nullptr, nullptr,
-                         &signed_squared_distance);
-
+                         &signed_squared_distance, &is_endpoint);
+  if (is_endpoint) {
```
Comment (dfridovi) [inline src/semiquadratic_polyline2_cost.cpp:61]: See earlier comment about whitespace.

File: src/polyline2.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (dfridovi) [reply src/polyline2.cpp:94]: actually you can just disregard this

File: src/polyline2_signed_distance_constraint.cpp (hunk 17 above)
Comment (dfridovi) [reply src/polyline2_signed_distance_constraint.cpp:68]: On second thought, just disregard this comment

File: src/quadratic_polyline2_cost.cpp (hunk 22 above)
Comment (dfridovi) [reply src/quadratic_polyline2_cost.cpp:66]: On second thought, you can just disregard this comment

File: src/semiquadratic_polyline2_cost.cpp (hunk 33 above)
Comment (dfridovi) [reply src/semiquadratic_polyline2_cost.cpp:61]: On second thought, just ignore this comment
